### PR TITLE
New driver for the push up method refactoring (P13) (retry 2)

### PR DESCRIPTION
**Pull request by carolahp**
### Summary
Implements a new driver for handling the user interaction logic when applying the **push up method refactoring**.
Tests are created for the new ReConditions.

The test `RBPullUpMethodParametrizedTest>>testPullUpReferencesInstVar` was modified because it pulled up a variable when the user replied "yes" to a question displayed in a dialog. In the current version the mechanics to pull up variables used in methods was removed from the Refactoring and put in the Driver.

### CODE REVIEW

**Comment by carolahp** [comment]
A single test is failing on windows, and it is unrelated to this PR

**Comment by Ducasse** [comment]
Oh this is cool!

**Comment by Ducasse** [comment]
I will get back to this probably tomorrow evening or tuesday evening.

**Comment by carolahp** [comment]
Ok, Thank you Stef

**Comment by Ducasse** [comment]
Hi caro @balsa-sarenac 
Balsa I need your eyes too. Caro I could not see a driver (may be I missed it) but I wonder
how we could avoid the confirm: in to the refactoring. A refactoring should not raise questions. It should be configured to be able to be used without or with interaction and
in general the interaction should be driven by its driver.

**Comment by Ducasse** [comment]
The confirm: are in RBPullUpMethodRefactoring >> checkInstVars

File: src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	#category : 'Refactoring-UI-Choices',
+	#package : 'Refactoring-UI',
+	#tag : 'Choices'
+}
+
+{ #category : 'execution' }
+ReRemoveDuplicateMethodsChoice >> action [
+
+	driver removeDuplicateMethods
+]
```
Comment (balsa-sarenac) [inline src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st:13]: This doesn't feel like a choice, but an additional option that can be done besides pull up. This doesn't have to be resolved in this PR, but I see these as checkboxes that user can tick and we do only those that user selected.

We can create an issue for this and resolve it in a separate PR.


These can be choices as well, but we have to combine them with other so we have 4 options and need to create at least 8 choice entries to cover all the cases, but I do not recommend this approach.

File: src/Refactoring-Core/RBPullUpMethodRefactoring.class.st [hunk 2]
```diff
@@ -127,18 +130,18 @@ RBPullUpMethodRefactoring >> checkInstVars [
...
+					[ (self confirm: ('<1p> refers to #<2s> which is defined in <3p>. Do you want pull up variable #<2s> also?' expandMacrosWith: aSelector
 								with: each
 								with: class))
-						ifTrue: [ self pushUpVariable: each ]
-						ifFalse: [ self refactoringWarning: 'You are about to push your method without the instance variable it uses.
+						ifTrue: [ self pullUpVariable: each ]
+						ifFalse: [ self refactoringWarning: 'You are about to pull your method without the instance variable it uses.
 						It will bring the system is an inconsistent state. But this may be what you want.
-						So do you want to push up anyway?' ] ]]
+						So do you want to pull up anyway?' ] ]]
```
Comment (balsa-sarenac) [inline src/Refactoring-Core/RBPullUpMethodRefactoring.class.st:144]: @Ducasse I think you were refering to these warnings that need to be moved to driver.
This PR is getting very long, I would also suggest to merge this once non-negotiables are fixed and then in a next PR fix this and options.

File: src/Refactoring-Core/RBPullUpMethodRefactoring.class.st [hunk 3]
```diff
@@ -282,8 +340,7 @@ RBPullUpMethodRefactoring >> privateTransform [
 	self
 		copyDownMethods;
 		pullUpMethods;
-		removePulledUpMethods;
-		removeDuplicateMethods
```
Comment (balsa-sarenac) [inline src/Refactoring-Core/RBPullUpMethodRefactoring.class.st:286]: Why was `removeDuplicateMethods` removed? Is it because it was moved to driver?

File: src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st [hunk 4]
```diff
@@ -263,19 +263,21 @@ RBPullUpMethodParametrizedTest >> testPullUpReferencesInstVar [
...
 		andArguments: { #(#foo ) . class }.
-	[ self executeRefactoring: refactoring ] valueSupplyingAnswer: true.
+	self proceedThroughWarning: [ self executeRefactoring: refactoring  ].
 	superClass := model classNamed: ('RBFooLint', 'RuleTestData1') asSymbol.
 	self assert: (superClass parseTreeForSelector: #foo)
 				equals: (self parseMethod: 'foo ^ foo').
-	self assert: (superClass directlyDefinesInstanceVariable: #foo).
 	self deny: (class directlyDefinesMethod: #foo).
-	self deny: (class directlyDefinesInstanceVariable: #foo)
+	self assert: (class directlyDefinesInstanceVariable: #foo)
```
Comment (balsa-sarenac) [inline src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st:271]: What is the reason for this change? I didn't see change in refactoring that would cause this? Just wanted to double-check that this wasn't a regression.

File: src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st [hunk 5]
```diff
@@ -263,19 +263,21 @@ RBPullUpMethodParametrizedTest >> testPullUpReferencesInstVar [
...
 { #category : 'failure tests' }
 RBPullUpMethodParametrizedTest >> testWarnDoesntPullUpReferencesInstVar [
 
 	| refactoring class |
+	rbClass = RBPullUpMethodTransformation ifTrue: [ " RBPullUpMethodTransformation shouldn't warn when instance variables are referenced "
+		self skip ].
+	
 	class := model classNamed:
 		         ('RBTransformation' , 'RuleTestData1') asSymbol.
 	refactoring := self
```
Comment (balsa-sarenac) [inline src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st:283]: This is a ok, but we were starting to separate tests for refactoring and transformations. Transformations tests are parametrized with refactorings as well, and refactorings only have their own tests that are concerned with breaking changes, so in a next PR it makes sense to move this to a new test class for refactoring.

File: src/Refactoring-Transformations/RBPullUpMethodTransformation.class.st [hunk 6]
```diff
@@ -291,26 +328,25 @@ RBPullUpMethodTransformation >> removeDuplicateMethods [
...
+			  ((aClass directlyDefinesMethod: aSelector) and: [
+				   (tree
+					    equalTo: (aClass parseTreeForSelector: aSelector)
+					    exceptForVariables: #(  )) and: [
+					   (aClass superclass whichClassIncludesSelector: aSelector)
+					   == targetSuperclass ] ]) ifTrue: [
+				  removeDuplicates ifFalse: [
+					  removeDuplicates := self refactoringConfirmWarning:
+						                      'Do you want to remove duplicate subclass methods?' ].
+				  removeDuplicates ifTrue: [
```
Comment (balsa-sarenac) [inline src/Refactoring-Transformations/RBPullUpMethodTransformation.class.st:345]: We want to remove all these warnings to driver from both transformation and refactoring. We can introduce flags for scripting purpose so user can configure it and we will set sane defaults.

**Review by balsa-sarenac — COMMENTED** [review]
Hey Caro, nice work, I see a lot of nice changes, but there is a lot of them (that's why it took me a while to review). In the future I would suggest more smaller PRs. For example, this could be at least two PRs: one for cleaning of conditions, their tests and usage in refactorings, and one for driver.
I don't think I saw tests for all the new conditions, can you confirm they are all tested?
As you will see below, I would suggest fixing only urgent things here like tests, and then finish this PR and continue in a new one that should be of a smaller size. I'm convinced I missed some things as well, so I will have to load the code and inspect when you do the next PR.

**Comment by carolahp** [comment]
> The confirm: are in RBPullUpMethodRefactoring >> checkInstVars

Hi Stef, you are right. Besides from `RBPullUpMethodRefactoring >> checkInstVars`, there are other methods whose name start with `check...` (both in `RBPullUpMethodRefactoring` and in `RBPullUpMethodTransformation`) which interact with the user. However they have no senders anymore. To avoid making this PR grow even more, I will create a new PR to remove these unused methods once we are done with this one

**Comment by Ducasse** [comment]
So do you suggest that I integrate it now. 
And balsa is right. Small PRs are better. And if you have a long PR please commits many small commits.

**Comment by Ducasse** [comment]
Caro can you check the feedback of balsa and fix the easy one and after we can integrate and iterate. 
Is it ok for you?

**Comment by carolahp** [comment]
> Caro can you check the feedback of balsa and fix the easy one and after we can integrate and iterate. Is it ok for you?

Hi Stef, yes, I'm on it. In some hours I will ping you to integrate

**Comment by carolahp** [comment]
> I don't think I saw tests for all the new conditions, can you confirm they are all tested?
Hi Balša. First of all, thanks a lot for the review! 

Regarding missing tests, you were right and I have just added tests for all the new conditions. 

In the following hours I will be replying to your comments one by one

File: src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st (hunk 1 above)
Comment (carolahp) [reply src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st:13]: I have currently implemented it using a checkbox as shown in this video. I am not sure this is what you meant in your comment. If not, I will create an issue and we can continue this discussion there,
RemoveDuplicateMethods

File: src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st (hunk 4 above)
Comment (carolahp) [reply src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st:271]: This modification was necessary because the refactoring is no longer pulling up references instance variables. The method responsible for pulling up referenced inst vars too was `checkInstVars`. Even though I didn't modify the method, I did remove the references to it from the refactoring. This can be seen in the images below.
image
image

File: src/Refactoring-Core/RBPullUpMethodRefactoring.class.st (hunk 3 above)
Comment (carolahp) [reply src/Refactoring-Core/RBPullUpMethodRefactoring.class.st:286]: Exactly, removeDuplicateMethods is now offered as a choice (as presented in the video above)

File: src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st (hunk 5 above)
Comment (carolahp) [reply src/Refactoring-Transformations-Tests/RBPullUpMethodParametrizedTest.class.st:283]: I created a new [issue](https://github.com/pharo-project/pharo/issues/17157) to move refactoring tests to a new class

File: src/Refactoring-Core/RBPullUpMethodRefactoring.class.st (hunk 2 above)
Comment (carolahp) [reply src/Refactoring-Core/RBPullUpMethodRefactoring.class.st:144]: I have created [issue#17158](https://github.com/pharo-project/pharo/issues/17158) to handle this

File: src/Refactoring-Transformations/RBPullUpMethodTransformation.class.st (hunk 6 above)
Comment (carolahp) [reply src/Refactoring-Transformations/RBPullUpMethodTransformation.class.st:345]: I modified [issue#17158](https://github.com/pharo-project/pharo/issues/17158) to handle this too

**Comment by carolahp** [comment]
@Ducasse we are ready to merge! I solved all @balsa-sarenac comments that I could do fast, and I created new issues for those who remain unresolved

File: src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	| myClassAlpha myClassBeta cond |
+	myClassAlpha := self model classNamed: #MyClassAlpha. 
+	myClassBeta := self model classNamed: #MyClassBeta.
+	
+	cond := ReMethodsSendNoSupersendsCondition new
+			class: myClassBeta
+		   targetSuperclass: myClassAlpha
+		   selectors: { #methodForPullUp }.
+		
+	" the method has a duplicate in MyClassBeta's sibling (MyClassBetaSibling), therefore the condition fails "
```
Comment (balsa-sarenac) [inline src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st:32]: I think this comment was left by accident @carolahp

File: src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	| myClassAlpha myClassBeta cond |
+	myClassAlpha := self model classNamed: #MyClassAlpha. 
+	myClassBeta := self model classNamed: #MyClassBeta.
+	
+	cond := ReMethodsSendNoSupersendsCondition new
+			class: myClassBeta
+		   targetSuperclass: myClassAlpha
+		   selectors: { #methodWithSuperSend }.
+		
+	" the method has a duplicate in MyClassBeta's sibling (MyClassBetaSibling), therefore the condition fails "
```
Comment (balsa-sarenac) [inline src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st:47]: This one too

File: src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st (hunk 1 above)
Comment (balsa-sarenac) [reply src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st:13]: I don't see a checkbox in the video, I will have to load the code and test it.

File: src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st (hunk 7 above)
Comment (carolahp) [reply src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st:32]: Thank you Balsa, I will remove it in the PR for the cleanup [issue#17158](https://github.com/pharo-project/pharo/issues/17158)

File: src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st (hunk 8 above)
Comment (carolahp) [reply src/Refactoring-Core-Tests/ReMethodsSendNoSupersendsTest.class.st:47]: Thanks, same as above

File: src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st (hunk 1 above)
Comment (carolahp) [reply src/Refactoring-UI/ReRemoveDuplicateMethodsChoice.class.st:13]: @balsa-sarenac Yes please, I would like to know what you think. Now that this PR is merged, you can make the checkbox appear by pulling up the method `MyClassBeta>>methodDuplicatedInSubclass`

**Comment by MarcusDenker** [comment]
This PR was merged with a failing test. I opendend am issue here:

https://github.com/pharo-project/pharo/issues/17175

Please, if a PR by accident is merged, we have to either fix it fast or revert. Else the CI will fail for every other PR checked, which in the past lead always to even more PRs with additional failing tests getting merged, leading to the CI degrading more and more.

**Comment by balsa-sarenac** [comment]
@carolahp could you do a really small PR with fixes to tests that are failing in the last CI run of this PR?
